### PR TITLE
Add gix diff tree

### PR DESCRIPTION
**Pull request by cruessler**
This PR adds a very basic `gix diff tree` subcommand. The output is very low level because it is supposed to help in debugging `gix blame`. Let me know if this is going in the right direction in which case I will replace the few `TODO` comments by proper ones.

### CODE REVIEW

File: gitoxide-core/src/repository/diff.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    out: &mut dyn std::io::Write,
+    old_treeish: BString,
+    new_treeish: BString,
+) -> anyhow::Result<()> {
+    let old_tree_id = repo.rev_parse_single(old_treeish.as_bstr())?;
+    let new_tree_id = repo.rev_parse_single(new_treeish.as_bstr())?;
+
+    let old_tree = old_tree_id
+        .object()
+        .unwrap()
```
Comment (Byron) [inline gitoxide-core/src/repository/diff.rs:14]: Unwraps aren't necessary here, thanks to `anyhow::Result`.

File: gitoxide-core/src/repository/diff.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        .into_tree();
+    let new_tree = new_tree_id
+        .object()
+        .unwrap()
+        .peel_to_kind(gix::object::Kind::Tree)
+        .unwrap()
+        .into_tree();
+
+    let changes = repo.diff_tree_to_tree(&old_tree, &new_tree, None)?;
```
Comment (Byron) [inline gitoxide-core/src/repository/diff.rs:25]: Nice to see you found this (new) method useful.

**Review by Byron — COMMENTED** [review]
Thanks for getting this started!

Please feel free to use any output style that you find useful, `gix` is a developer tool and thus is there to be useful to you, not to replicate `git`.

Once the basic documentation of the CLI is done, merging this PR won't take long.

File: gitoxide-core/src/repository/diff.rs (hunk 2 above)
Comment (cruessler) [reply gitoxide-core/src/repository/diff.rs:25]: It was when I saw it in the history that I decided to open this PR. :-)

**Comment by cruessler** [comment]
Perfect! I made a few minor changes, so from my side it’s good to go. I hope I’ve picked up the right patterns with respect to documentation, wording and style, let me know if I haven’t!

**Review by Byron — APPROVED** [review]
Thanks, this looks great!

I altered the output a bit to be more compact, which should be easier to parse for humans:

```
M: gitoxide-core
  07d9a8eaca88ef0ed1fc1b62867a699519f72918 -> 4e07fa7a1ee0fee4a02674ea983f38a2432c029e
  40000 -> 40000
M: src
  d4694a7c701e721067dc93fde402dea6e93c2202 -> 34a16f1a9422e3aee256d07fba533615b2f11b98
  40000 -> 40000
M: gitoxide-core/src
  9d54cff8220f8954cf751dc0645eff626b36c1ff -> ff3058350b64f4557659a3342dd92d3b5c5b9314
  40000 -> 40000
M: src/plumbing
  5fd37efa8f8b130ddb882a880442a4807c0e16b3 -> 9dabb47090b0036f55fa84c8a34cf176178b186f
  40000 -> 40000
M: gitoxide-core/src/repository
  33a974f21369bdc7b75eb6b2c83de3b18e252a80 -> 0e017b40f7398de02dc0e7446cd13c49718b755b
  40000 -> 40000
M: src/plumbing/options
  87f1da1a38158669b566fd8c4da9899f85e1dd93 -> 670c2694f21797f29ce29d435a1e0c9a3398e201
  40000 -> 40000
M: gitoxide-core/src/repository/diff.rs
  f6b515682076d1dcd87772536b3b5f560a57a9bd -> 18ca60f9e88f2694651f6202c49882520d76d468
  100644 -> 100644
M: src/plumbing/options/mod.rs
  b038d3ca0916e5c0c7a784f37edbbc3481a131b7 -> b6b1b005f110e92403cd5a74f0b9e552f662596d
  100644 -> 100644
```

…is now…

```
M: gitoxide-core/
  07d9a8eac -> 4e07fa7a1
M: src/
  d4694a7c7 -> 34a16f1a9
M: gitoxide-core/src/
  9d54cff82 -> ff3058350
M: src/plumbing/
  5fd37efa8 -> 9dabb4709
M: gitoxide-core/src/repository/
  33a974f21 -> 0e017b40f
M: src/plumbing/options/
  87f1da1a3 -> 670c2694f
M: gitoxide-core/src/repository/diff.rs
  f6b515682 -> 18ca60f9e
M: src/plumbing/options/mod.rs
  b038d3ca0 -> b6b1b005f
```

It's possible that you wanted it like it was for other reasons, and if the case I think you could add another PR with an option that reverts it to the previous rendering style. Or maybe, just revert it to what you need and it can stay like this until `gix blame` is fully functional.

**Comment by Byron** [comment]
And…I actually found something I will investigate :D. Great tool already!

**Comment by cruessler** [comment]
> Thanks, this looks great!
> 
> I altered the output a bit to be more compact, which should be easier to parse for humans:
> 
> ```
> M: gitoxide-core
>   07d9a8eaca88ef0ed1fc1b62867a699519f72918 -> 4e07fa7a1ee0fee4a02674ea983f38a2432c029e
>   40000 -> 40000
> M: src
>   d4694a7c701e721067dc93fde402dea6e93c2202 -> 34a16f1a9422e3aee256d07fba533615b2f11b98
>   40000 -> 40000
> M: gitoxide-core/src
>   9d54cff8220f8954cf751dc0645eff626b36c1ff -> ff3058350b64f4557659a3342dd92d3b5c5b9314
>   40000 -> 40000
> M: src/plumbing
>   5fd37efa8f8b130ddb882a880442a4807c0e16b3 -> 9dabb47090b0036f55fa84c8a34cf176178b186f
>   40000 -> 40000
> M: gitoxide-core/src/repository
>   33a974f21369bdc7b75eb6b2c83de3b18e252a80 -> 0e017b40f7398de02dc0e7446cd13c49718b755b
>   40000 -> 40000
> M: src/plumbing/options
>   87f1da1a38158669b566fd8c4da9899f85e1dd93 -> 670c2694f21797f29ce29d435a1e0c9a3398e201
>   40000 -> 40000
> M: gitoxide-core/src/repository/diff.rs
>   f6b515682076d1dcd87772536b3b5f560a57a9bd -> 18ca60f9e88f2694651f6202c49882520d76d468
>   100644 -> 100644
> M: src/plumbing/options/mod.rs
>   b038d3ca0916e5c0c7a784f37edbbc3481a131b7 -> b6b1b005f110e92403cd5a74f0b9e552f662596d
>   100644 -> 100644
> ```
> 
> …is now…
> 
> ```
> M: gitoxide-core/
>   07d9a8eac -> 4e07fa7a1
> M: src/
>   d4694a7c7 -> 34a16f1a9
> M: gitoxide-core/src/
>   9d54cff82 -> ff3058350
> M: src/plumbing/
>   5fd37efa8 -> 9dabb4709
> M: gitoxide-core/src/repository/
>   33a974f21 -> 0e017b40f
> M: src/plumbing/options/
>   87f1da1a3 -> 670c2694f
> M: gitoxide-core/src/repository/diff.rs
>   f6b515682 -> 18ca60f9e
> M: src/plumbing/options/mod.rs
>   b038d3ca0 -> b6b1b005f
> ```
> 
> It's possible that you wanted it like it was for other reasons, and if the case I think you could add another PR with an option that reverts it to the previous rendering style. Or maybe, just revert it to what you need and it can stay like this until `gix blame` is fully functional.

Perfect! I had thought about making the output more condensed, but was unsure. Seeing your changes, I now know that I want the output more compact as well. :-)

**Pull request merged**